firmware/shared/code/I2C.c

```diff
@@ -0,0 +1,268 @@
+	NVIC_Init_Struct.NVIC_IRQChannelPreemptionPriority = 5;
+	NVIC_Init_Struct.NVIC_IRQChannelSubPriority        = 0;
+	NVIC_Init(&NVIC_Init_Struct);
+
+	// NVIC Init for error interrupt
+	NVIC_Init_Struct.NVIC_IRQChannel                   = I2C2_ER_IRQn;
+	NVIC_Init_Struct.NVIC_IRQChannelCmd                = ENABLE;
+	NVIC_Init_Struct.NVIC_IRQChannelPreemptionPriority = 4;
+	NVIC_Init_Struct.NVIC_IRQChannelSubPriority        = 0;
+	NVIC_Init(&NVIC_Init_Struct);
+
+	// Set priority for ISR and enable ISR
+	//NVIC_SetPriority(I2C2_EV_IRQn, 0);
+	NVIC_EnableIRQ(I2C2_EV_IRQn);
+	NVIC_EnableIRQ(I2C2_ER_IRQn);
+}
+
+void I2C_setup(void)
+{
+	// I2C2 Init
+	RCC_APB1PeriphClockCmd(RCC_APB1Periph_I2C2, ENABLE);
+
+	/*
+	 * Reset I2C after a delay: After enabling the RCC clock, the BUSY bit in the SR2 register
+	 *                          became 1, which means the peripheral cannot start any data
+	 *                          transaction. According to STM32 documents and resources on
+	 *                          Google, this could be a silicon defect, and reseting the RCC
+	 *                          clock after a small delay is the current work-around.
+	 * (https://community.st.com/s/question/0D50X00009XkYxY/stm32f105rc-i2c2-always-busy)
+	 *
+	 * 15 is the minimum value of delay that ensures I2C's functioning for now.
+	 * If the BUSY bit cannot be reset, increase the delay time.
+	 * DO NOT check BUSY bit in SR2 in the code!!! It will then be set FOREVER!!!
+	 */
+	int i = 0;
+	while (i < 15){
+		i++;
+	}
+	RCC_APB1PeriphResetCmd(RCC_APB1Periph_I2C2, ENABLE);
+	RCC_APB1PeriphResetCmd(RCC_APB1Periph_I2C2, DISABLE);
+
+	I2C_InitTypeDef I2C_Init_Struct;
+	I2C_Init_Struct.I2C_ClockSpeed          = 50000; // Any requirement for I2C clock speed?
+	I2C_Init_Struct.I2C_Ack                 = I2C_Ack_Enable;
+	I2C_Init_Struct.I2C_AcknowledgedAddress = I2C_AcknowledgedAddress_7bit;
+	I2C_Init_Struct.I2C_DutyCycle           = I2C_DutyCycle_2;
+	I2C_Init_Struct.I2C_Mode                = I2C_Mode_I2C;
+	I2C_Init_Struct.I2C_OwnAddress1         = 0x0;
+	I2C_Init(I2C2, &I2C_Init_Struct);
+
+	// Enable event interrupts and buffer interrupts
+	I2C_ITConfig(I2C2, I2C_IT_EVT, ENABLE);
+	I2C_ITConfig(I2C2, I2C_IT_BUF, ENABLE);
+	I2C_ITConfig(I2C2, I2C_IT_ERR, ENABLE);
+
+	// Enable I2C2
+	I2C_Cmd(I2C2, ENABLE);
+}
+
+bool I2C_send(I2C_channel_E channel, const uint8_t * data, const uint8_t length)
+{
+	// Check the validity of message
+	if((data != NULL) && (length != 0) && (length <= _I2C_BUFFER_SIZE))
+	{
+		// Check the availability of I2C
+		if ((data_length != 0) || (I2C_state != I2C_STATE_IDLE)) return false;
+
+		memcpy(I2C_send_data.Buffer, data, length);
+		I2C_state = I2C_STATE_SEND;
+		slave_address = I2C_channelToAddressMapping(channel);
+		data_length = length;
+		I2C_GenerateSTART(I2C2, ENABLE);
+		return true;
+	}
+	else return false;
+}
+
+bool I2C_receive(I2C_channel_E channel, const uint8_t * data, const uint8_t length)

```

Comment:
Future work: Let's improve this receive using callbacks. There are some problems you'll encounter by passing in a pointer to a buffer and not populating the buffer right away. We have to be careful not to put the buffer out of scope. A callback can tell us exactly when the data arrives.